source/grid/grid_tools.cc

```diff
@@ -1716,6 +1716,209 @@ namespace GridTools
+
+    std::vector<Point<spacedim> > subdomain_boundary_cells_centers; // cache all the subdomain boundary cells centers here
+    std::vector<double> subdomain_boundary_cells_radii; // cache all the subdomain boundary cells radii
+    subdomain_boundary_cells_centers.reserve (subdomain_boundary_cells.size());
+    subdomain_boundary_cells_radii.reserve (subdomain_boundary_cells.size());
+    // compute cell radius for each boundary cell of the predicate subdomain
+    for ( typename std::vector<typename MeshType::active_cell_iterator>::const_iterator
+          subdomain_boundary_cell_iterator  = subdomain_boundary_cells.begin();
+          subdomain_boundary_cell_iterator != subdomain_boundary_cells.end(); ++subdomain_boundary_cell_iterator )
+      {
+        const std::pair<Point<spacedim>, double> &
+        subdomain_boundary_cell_enclosing_ball = (*subdomain_boundary_cell_iterator)->enclosing_ball();
+
+        subdomain_boundary_cells_centers.push_back( subdomain_boundary_cell_enclosing_ball.first);
+        subdomain_boundary_cells_radii.push_back( subdomain_boundary_cell_enclosing_ball.second);
+      }
+    AssertThrow( subdomain_boundary_cells_radii.size() == subdomain_boundary_cells_centers.size(),
+                 ExcInternalError());
+
+    // Find the cells that are within layer_thickness of predicate subdomain boundary
+    // distance but are inside the extended bounding box.
+    // Most cells might be outside the extended bounding box, so we could skip them.
+    // Those cells that are inside the extended bounding box but are not part of the
+    // predicate subdomain are possible candidates to be within the distance to the
+    // boundary cells of the predicate subdomain.
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      {
+        // Ignore all the cells that are in the predicate subdomain
+        if ( predicate(cell))
+          continue;
+
+        const std::pair<Point<spacedim>, double> &cell_enclosing_ball
+          = cell->enclosing_ball();
+
+        const Point<spacedim> &cell_enclosing_ball_center = cell_enclosing_ball.first;
+        const double &cell_enclosing_ball_radius = cell_enclosing_ball.second;
+
+        bool cell_inside = true; // reset for each cell
+
+        for (unsigned int d = 0; d < spacedim; ++d)
+          cell_inside &= (cell_enclosing_ball_center[d] + cell_enclosing_ball_radius > bounding_box.first[d])
+                         && (cell_enclosing_ball_center[d] - cell_enclosing_ball_radius < bounding_box.second[d]);
+        // cell_inside is true if any of the cell vertices are inside the extended bounding box
+
+        // Ignore all the cells that are outside the extended bounding box
+        if (cell_inside) // cell is
+          for (unsigned int i =0; i< subdomain_boundary_cells_radii.size(); ++i)
+            if ( (cell_enclosing_ball_center - subdomain_boundary_cells_centers[i]).norm_square()
```

Comment:
you can avoid temporary `Tensor<1,spacedim>` object here by doing `cell_enclosing_ball_center. distance_square(subdomain_boundary_cells_centers[i])`